api/src/license_operator.rs

```diff
@@ -0,0 +1,207 @@
+use std::borrow::Borrow;
+
+use log::info;
+use rand::rngs::ThreadRng;
+use rand::Rng;
+use sha3::{
+    digest::{ExtendableOutput, Update, XofReader},
+    Shake256,
+};
+
+use crate::adler32::adler32_checksum;
+use crate::license_key::LicenseKeyStatus;
+use crate::license_magic::LicenseMagic;
+use crate::license_serializer::{DefaultLicenseKeySerializer, LicenseKeySerializer};
+use crate::LicenseKey;
+
+pub struct LicenseOperator {
+    rng: ThreadRng,
+
+    magic: LicenseMagic,
+    key_size: usize,
+    checksum_init: (u32, u32),
+
+    serializer: Box<dyn LicenseKeySerializer>,
+
+    blacklist: Vec<Vec<u8>>,
+    byte_check: Vec<(usize, LicenseMagic)>,
+}
+
+impl LicenseOperator {
+    #[inline(always)]
+    pub fn new(
+        magic: LicenseMagic,
+        key_size: usize,
+        checksum_init: (u32, u32),
+        serializer: Box<dyn LicenseKeySerializer>,
+        blacklist: Vec<Vec<u8>>,
+        byte_check: Vec<(usize, LicenseMagic)>,
+    ) -> LicenseOperator {
+        LicenseOperator {
+            rng: rand::thread_rng(),
+            magic,
+            key_size,
+            checksum_init,
+            serializer,
+            blacklist,
+            byte_check,
+        }
+    }
+
+    #[inline(always)]
+    pub fn default() -> LicenseOperator {
+        LicenseOperator {
+            rng: rand::thread_rng(),
+            magic: LicenseMagic::default(),
+            key_size: 16,
+            checksum_init: (0xFFAA, 0xAAFF),
+            serializer: Box::new(DefaultLicenseKeySerializer {}),
+            blacklist: Vec::new(),
+            byte_check: Vec::new(),
+        }
+    }
+
+    #[inline(always)]
+    pub fn randomize_magic(&mut self, magic_size: usize, magic_count: usize) -> &Self {
+        for _x in 0..magic_size {
```

Comment:
This is not really best way how to do for loops in rust :slightly_smiling_face: . They does not work the same way asi in c++, they are closer to for iterators. Also ThreadRng has method for [filling da bytes](https://rust-random.github.io/rand/rand/trait.RngCore.html#tymethod.fill_bytes) and also you know the size of vector beforehand, so you can actually allocate it accordingly using `.with_capacity()` Here is an example how I would tackle such problem:
```rust
pub fn randomize_magic(&mut self, magic_size: usize, magick_count: usize) -> &Self {
    let mut rng = rand::thread_rng();
    let gen = &mut rng  // https://stackoverflow.com/questions/33482764/how-to-capture-mutable-reference-into-move-closure-contained-in-iterator-returne
    self.magic.extend_from_slizce((0..magic_count).map(|_| {
        let mut buf_magic = Vec::<u8>::with_capacity(magic_size);
        gen.fill_bytes(&mut buf_magic);
        buf_magic
    }).into())  // or .collect::<Vec<Vec<u8>>().as_ref(), or & etc..
}
```

However do you think it is a good idea to randomize magic, when you have already setup `LicenseOperator`. Wouldn't something like builder be more appropriate, or some kind of factory funciton?

Reply:
Well I simplified with more idiomatic way.

```rust
pub fn randomize_magic(&mut self, magic_size: usize, magic_count: usize) {
        let mut rng = rand::thread_rng();

        for _ in 0..magic_size {
            self.magic
                .push((0..magic_count).map(|_| rng.gen()).collect());
        }
    }
```

> However do you think it is a good idea to randomize magic, when you have already setup LicenseOperator. Wouldn't something like builder be more appropriate, or some kind of factory funciton?

Oh... yeah I should call it in "builder", what I also should do, is implement writing those magic number to some file, user/devs will definitely need that number

Reply:
Don't know. Doesn't it complain about move of the rng? Or does it create internally some mutable ref? Because you may also create just bunch of `Map` operators that have reference to `rng` and send them over to another thread, or out of scope. If you really want keep that for iterator, isn't better way to initialize that thread local random nuber gen inside that for loop?

```rust
for _ in 0..magic_cout {
    let mut rng = rand::thread_rng();  // this way you can spawn it in parallel using e.g. rayon

    let mut buffer = Vec::<u8>::with_capacity(magic_size);

    rng.fill_bytes(&mut buffer);

    self.magic.push(buffer);
}
```

The case using rayon should then look like:
```rust
use rayon::prelude::*;


pub fn generate_magic(length: usize, count: usize) -> Vec<Vec<u8>> {
    (0..count).par_iter().map(|_| {
        let mut rng = rand::thread_rng();
        // create Vec using rng
    }).collect()
}
```

Reply:
Well rng and compiler are happy about my solution. I actually planned to use rayon but... in some future, right now I just want functional library